test/Test_TypeAssignment.re

```diff
@@ -0,0 +1,603 @@
+    uexp_fun(name, ty1, uexp_fun_gen(Ctx.extend(ctx, entry), ty2, n));
+  | _ => uexp_gen(ctx, ty, n / 2)
+  };
+}
+
+//Generate an ap expression with bound n
+and uexp_ap_gen = (ctx: Ctx.t, ty: UTyp.t, fn: UExp.t, n: int): UExp.t => {
+  switch (ty.term) {
+  | Arrow(ty1, ty2) =>
+    let arg = uexp_gen(ctx, ty1, n / 2);
+    uexp_ap_gen(ctx, ty2, uexp(Ap(fn, arg)), n);
+  | _ => fn
+  };
+}
+
+//Generate match expression with bound n
+and uexp_match_gen = (ctx: Ctx.t, ty: UTyp.t, n: int): UExp.t => {
+  let scrut_ty = utyp_match_gen |> gen1;
+  let scrut = uexp_gen(ctx, scrut_ty, n / 2);
+
+  let length = int_range(1, 4) |> gen1;
+  let rec uexp_match_helper =
+          (length: int, acc: list((UPat.t, UExp.t)))
+          : list((UPat.t, UExp.t)) => {
+    switch (length) {
+    | 0 => acc
+    | m =>
+      let (entries, p) = upat_gen([], scrut_ty, n / 2);
+      let ctx' =
+        List.fold_left(
+          (ctx, entry) => Ctx.extend(ctx, entry),
+          ctx,
+          entries,
+        );
+      let u = uexp_gen(ctx', ty, n / 2);
+      uexp_match_helper(m - 1, [(p, u), ...acc]);
+    };
+  };
+  let l = uexp_match_helper(length, []);
+  Match(scrut, l) |> uexp;
+}
+
+//Generates a UExp with bound n
+//if n is 0, then generate a base expression
+and uexp_gen = (ctx: Ctx.t, ty: UTyp.t, n: int): UExp.t => {
+  let rule = n == 0 ? Base : rules |> gen1;
+  switch (rule) {
+  | Base => uexp_base_gen(ctx, ty)
+  | Ap =>
+    let fn_typ = utyp_arrow_gen(ty) |> gen1;
+    let fn = uexp_fun_gen(ctx, fn_typ, n / 2);
+    uexp_ap_gen(ctx, fn_typ, fn, n / 2);
+  | If =>
+    If(
+      uexp_gen(ctx, utyp(Bool), n / 2),
+      uexp_gen(ctx, ty, n / 2),
+      uexp_gen(ctx, ty, n / 2),
+    )
+    |> uexp
+  | Match => uexp_match_gen(ctx, ty, n / 2)
+  };
+};
+
+let pure_random_utyp: QCheck.Gen.t(UTyp.t) =
+  QCheck.Gen.sized_size(
+    int_range(0, 1),
+    fix((self, n) =>
+      switch (n) {
+      | 0 =>
+        frequency([
+          (1, pure(utyp(Int))),
+          (1, pure(utyp(Bool))),
+          (1, pure(utyp(Float))),
+          (1, pure(utyp(String))),
+        ])
+      | n =>
+        frequency([
+          (1, map2(utyp_arrow, self(n / 2), self(n / 2))),
+          (1, map(utyp_list, self(n / 2))),
+        ])
+      }
+    ),
+  );
+
+let pure_random_upat: QCheck.Gen.t(UPat.t) =
+  QCheck.Gen.sized_size(
+    int_range(1, 2),
+    fix((self, n) =>
+      switch (n) {
+      | 0 =>
+        frequency([
+          (1, map(upat_int, int)),
+          (1, map(upat_bool, bool)),
+          (1, map(upat_float, float)),
+          (1, map(upat_string, printable_string)),
+          (1, map(upat_var, printable_string)),
+        ])
+      | n =>
+        frequency([
+          (1, self(n / 2)),
+          (1, map(upat_list, list_size(int_range(1, 2), self(n / 2)))),
+        ])
+      }
+    ),
+  );
+
+//Generates a UExp with no regard to the typing rules
+let pure_random_uexp: QCheck.Gen.t(UExp.t) =
```

Comment:
We should try to build a full arbitrary instance for the types so we can add printers/shrinking functions in the future like https://github.com/c-cube/qcheck?tab=readme-ov-file#mirrors-and-trees

Reply:
I feel like we should be able to use the ppx https://github.com/c-cube/qcheck/tree/main/src/ppx_deriving_qcheck for at least the simple variants.